aten/src/ATen/native/RNN.cpp

```diff
@@ -420,7 +480,95 @@ struct LSTMCell : Cell<std::tuple<Tensor, Tensor>, cell_params> {
+    const auto _input = pre_compute_input ? input : params.linear_ih(input);
+    const auto h_w_hh = params.linear_hh(hx);  // qint8
+    const auto i_w_ih = _input.expand({hx.size(0), -1});  // qint8
+    const auto gates = elementwise_arithmetic("quantized::add_out",
+                                              i_w_ih, h_w_hh,
+                                              kQ8BitScaleWithRange16,
+                                              kQZeroPoint);  // qint8
+
+    auto chunked_gates = gates.chunk(4, 1);
+    auto ingate = at::sigmoid(chunked_gates[0]);
+    auto forgetgate = at::sigmoid(chunked_gates[1]);
+    auto cellgate = at::tanh(chunked_gates[2]);
+    auto outgate = at::sigmoid(chunked_gates[3]);
+
+    auto in_cell = elementwise_arithmetic("quantized::mul_out",
+                                          ingate, cellgate,
+                                          kQ8BitScaleWithRange2,
+                                          kQZeroPoint);
+    auto d_in_cell = in_cell.dequantize();
+    auto d_forgetgate = forgetgate.dequantize();
+    auto d_cx = cx.dequantize();
+
+    auto f_cy = (d_forgetgate * d_cx) + d_in_cell;
+    auto cy = at::quantize_per_tensor(f_cy, kQ32BitScaleWithRange256,
+                                      kQZeroPoint,
+                                      at::kQInt32);
+    auto hy = elementwise_arithmetic("quantized::mul_out",
+                                     outgate, at::tanh(cy),
+                                     kQ8BitScaleWithRange2,
+                                     kQZeroPoint);
+    return std::make_tuple(std::move(hy), std::move(cy));
+  }
+
+private:
+  Tensor elementwise_arithmetic(const char* function_name,
+                                Tensor qa, Tensor qb,
+                                Scalar out_scale, Scalar out_zero_point) const {
+    static constexpr ScalarType qtype = at::kQInt8;
+    auto rqa = demoting_requantization(qa, out_scale, out_zero_point, qtype);
+    auto rqb = demoting_requantization(qb, out_scale, out_zero_point, qtype);
+
+    Tensor out = at::_empty_affine_quantized(rqa.sizes(),
+      at::device(at::kCPU).dtype(qtype),
+                 out_scale.toDouble(),
+                 out_zero_point.toLong());
+    callOp(function_name, "", rqa, rqb, out);
+    return out;
+  }
+
+  // Demotes from qint32 to qint8
+  Tensor demoting_requantization(Tensor qa, Scalar out_scale,
+                                 Scalar out_zero_point, ScalarType qtype) const {
+    // TODO: Need to optimize this part.
+    if (qa.scalar_type() == at::kQUInt8) {
+      auto dqa = qa.dequantize();
+      auto qqa = at::quantize_per_tensor(dqa, out_scale.toDouble(), out_zero_point.toLong(), qtype);
+      return qqa;
+    }
+    if (qa.scalar_type() != at::kQInt32) {
+      return qa;
+    }
+    auto dqa = qa.dequantize();
+    auto qqa = at::quantize_per_tensor(dqa, out_scale.toDouble(), out_zero_point.toLong(), qtype);
+    return qqa;
```

Comment:
Can we organize this control flow as
```
if (qa.scalar_type() == at::kQUInt8) {
  ...
} else if (qa.scalar_type() == at::kQInt32) {
  ...
} else {
  ...
}
```
or even as a `switch`.

It took me a second to reason about what's going on here